akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/MdcLoggingDirectivesSpec.scala

```diff
@@ -0,0 +1,95 @@
+package akka.http.scaladsl.server.directives
+
+import akka.event._
+import akka.http.scaladsl.server.RoutingSpec
+import akka.testkit.EventFilter
+
+class MdcLoggingDirectivesSpec extends RoutingSpec {
+
+  "The `withMarkerLoggingAdapter` directive" should {
+    "set a DiagnosticMarkerBusLoggingAdapter on the request context and provide the same adapter to the caller" in {
+      Get() ~> withMarkerLoggingAdapter { provided: MarkerLoggingAdapter =>
+        extractLog { extracted: LoggingAdapter =>
+          extracted shouldEqual provided
+          completeOk
+        }
+      } ~> check {
+        response shouldEqual Ok
+      }
+    }
+    "provides a new DiagnosticMarkerBusLoggingAdapter for each invocation" in {
+      Get() ~> withMarkerLoggingAdapter { _ =>
+        extractLog { log1 =>
+          withMarkerLoggingAdapter { _ =>
+            extractLog { log2 =>
+              log1 should not equal log2
+              completeOk
+            }
+          }
+        }
+      } ~> check {
+        response shouldEqual Ok
+      }
+    }
+  }
+
+  "The `withMdcEntries` and `withMdcEntry` directives" should {
+    "incrementally append entries to the LoggingAdapter MDC maps" in {
+      Get() ~> extractLog { log1 =>
+        withMdcEntry("foo", "foo entry 1") {
+          extractLog { log2 =>
+            withMdcEntries("foo" -> "foo entry 2", "bar" -> "bar entry 1") {
+              extractLog { log3 =>
+                log1.mdc shouldBe Map.empty
+                log2.mdc shouldBe Map("foo" -> "foo entry 1")
+                log3.mdc shouldBe Map("foo" -> "foo entry 2", "bar" -> "bar entry 1")
+                completeOk
+              }
+            }
+          }
+        }
+      } ~> check {
+        response shouldEqual Ok
+      }
+    }
+    "include the entries in the LoggingEvents" in {
+      class EF(n: Int) extends EventFilter(n) {
+        @volatile
+        var buf = Vector.empty[Logging.Info2]
+        override protected def matches(event: Logging.LogEvent): Boolean = event match {
+          case e: Logging.Info2 =>
+            buf :+= e
+            true
+          case _ => false
+        }
+      }
+      val filter = new EF(3)
```

Comment:
I'd recommend nesting 3 `EventFilter.custom` instead of the volatile field scheme, looking at the mdc values in the filters instead of after the fact. First log event will go innermost. I think it can match on exact log message and fail explicitly with a scalatest `fail` if the mdc isn't what it should be to provide extra detail.